File: scenes/screensavers/lines/point3d.py
import math

class Point3D() :
	def __init__(self, x, y, z, fl=250) :
		self._x = x
		self._y = y
		self._z = z
		self._fl = fl
		self._color = (255,255,255)
	
	def setVanishingPoint(self, vpX, vpY) :
		self._vpX = vpX
		self._vpY = vpY

	def setCenter(self, cX, cY, cZ) :
		self._cX = cX
		self._cY = cY
		self._cZ = cZ
	
	def rotateX(self, angleX) :
		
		cosX = math.cos(angleX)
		sinX = math.sin(angleX)
		
		y1 = self._y * cosX - self._z * sinX
		z1 = self._z * cosX + self._y * sinX
		
		self._y = y1
		self._z = z1

	def rotateY(self, angleY) :
		cosY = math.cos(angleY)
		sinY = math.sin(angleY)
		
		x1 = self._x * cosY - self._z * sinY
		z1 = self._z * cosY + self._x * sinY
		
		self._x = x1
		self._z = z1
	
	def rotateZ(self, angleZ) :
		cosZ = math.cos(angleZ)
		sinZ = math.sin(angleZ)
		
		x1 = self._x * cosZ - self._y * sinZ
		y1 = self._y * cosZ + self._x * sinZ
		
		self._x = x1
		self._y = y1

	@property
	def screenX(self) :
		scale = self._fl / (self._fl + self._z + self._cZ)
		return self._vpX + (self._cX + self._x) * scale

	@property
	def screenY(self) :
		scale = self._fl / (self._fl + self._z + self._cZ)
		return self._vpY + (self._cY + self._y) * scale
	
	@property
	def x(self) :
		return self._x

	@property
	def color(self) :
		return self._color

	@x.setter
	def x(self, x) :
		self._x = x
	
	@property
	def y(self) :
		return self._y

	@y.setter
	def y(self, y) :
		self._y = y

	@property
	def z(self) :
		return self._z

	@z.setter
	def z(self, z) :
		self._z = z
	
	@color.setter
	def color(self, color) :
		self._color = color

File: web/client/src/js/lib/point3d.js
export default class Point3d {
	constructor(x, y, z) {
		this.x = (x === undefined) ? 0 : x;
		this.y = (y === undefined) ? 0 : y;
		this.z = (z === undefined) ? 0 : z;
		this.fl = 250;
		this.vpX = 0;
		this.vpY = 0;
		this.cX = 0;
		this.cY = 0;
		this.cZ = 0;
	}

	setVanishingPoint(vpX, vpY) {
		this.vpX = vpX;
		this.vpY = vpY;
	}

	setCenter(cX, cY, cZ) {
		this.cX = cX;
		this.cY = cY;
		this.cZ = cZ;
	}

	rotateX(angleX) {
		var cosX = Math.cos(angleX),
			sinX = Math.sin(angleX),
			y1 = this.y * cosX - this.z * sinX,
			z1 = this.z * cosX + this.y * sinX;
		this.y = y1;
		this.z = z1;
	};

	rotateY(angleY) {
		var cosY = Math.cos(angleY),
			sinY = Math.sin(angleY),
			x1 = this.x * cosY - this.z * sinY,
			z1 = this.z * cosY + this.x * sinY;
		this.x = x1;
		this.z = z1;
	}

	rotateZ(angleZ) {
		var cosZ = Math.cos(angleZ),
			sinZ = Math.sin(angleZ),
			x1 = this.x * cosZ - this.y * sinZ,
			y1 = this.y * cosZ + this.x * sinZ;
		this.x = x1;
		this.y = y1;
	}

	getScreenX() {
		var scale = this.fl / (this.fl + this.z + this.cZ);
		return this.vpX + (this.cX + this.x) * scale;
	}

	getScreenY() {
		var scale = this.fl / (this.fl + this.z + this.cZ);
		return this.vpY + (this.cY + this.y) * scale;
	}
}
